modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerRequest.java

```diff
@@ -278,6 +298,53 @@ public RasterLayerRequest(
         spatialRequestHelper.compute();
     }
 
+    private void checkAlternativeCRSIsSupported() throws IOException {
+        if (heterogeneousCRS && useAlternativeCRS && requestedBounds != null) {
+            CoordinateReferenceSystem requestedCRS = requestedBounds.getCoordinateReferenceSystem();
+            String requestedEpsgCode = Utils.getEpsgCodeNumber(requestedCRS);
+            // Enable alternative CRS Output support only when the requested CRS doesn't match
+            // the coverage's one. In that case, proceed with the standard approach
+            if (rasterManager.hasAlternativeCRS(requestedEpsgCode)
+                    && !CRS.equalsIgnoreMetadata(
+                            requestedCRS, spatialRequestHelper.getReferenceCRS(false))) {
+                // Initialize the alternativeCRS Output Coverage properties
+                spatialRequestHelper.setSupportingAlternativeCRSOutput(true);
+                CoverageProperties alternativeProperties = new CoverageProperties();
+                alternativeProperties.setCrs2D(requestedCRS);
+                alternativeProperties.setBBox(ReferencedEnvelope.reference(requestedBounds));
+                spatialRequestHelper.setAlternativeProperties(alternativeProperties);
+            } else {
+                // We can't support the alternative CRS so let's disable it
+                useAlternativeCRS = false;
+            }
+        }
+    }
+
+    private void computeRequestedGridGeometry() throws IOException {
+        if (requestedGridGeometry != null) {
+            if (heterogeneousCRS && !useAlternativeCRS) {
+                GridEnvelope2D paddedRange =
+                        new GridEnvelope2D(requestedGridGeometry.getGridRange2D());
+                paddedRange.setBounds(
+                        paddedRange.x - DEFAULT_PADDING,
+                        paddedRange.y - DEFAULT_PADDING,
+                        paddedRange.width + DEFAULT_PADDING * 2,
+                        paddedRange.height + DEFAULT_PADDING * 2);
+
+                GridGeometry2D padded =
+                        new GridGeometry2D(
+                                paddedRange,
+                                requestedGridGeometry.getGridToCRS(),
+                                requestedGridGeometry.getCoordinateReferenceSystem());
+                spatialRequestHelper.setRequestedGridGeometry(padded.toCanonical());
+            } else {
+                // Do not apply any padding if we are going to produce output in requested CRS.
```

Comment:
Hum... say we have a bunch of granules and some are in the requested CRS, others are not. Is the padding going to be applied to the ones that are getting reprojected?